File: scripts/translate-grepreader-keys.mjs
// One-time script to translate GrepTag keys for all 10 locales using Claude API
import Anthropic from '@anthropic-ai/sdk';
import { readFileSync, writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const localesDir = join(__dirname, '..', 'src', 'locales');
const client = new Anthropic();

const LANGS = {
  es: 'Spanish', fr: 'French', nl: 'Dutch', pt: 'Portuguese',
  ru: 'Russian', zh: 'Simplified Chinese', ja: 'Japanese',
  ko: 'Korean', id: 'Indonesian', ms: 'Malay'
};

// Keys to translate — extracted from en locale
const en = JSON.parse(readFileSync(join(localesDir, 'en', 'translation.json'), 'utf8'));

const keysToTranslate = {
  grepreader: en.redesign.productFamily.products.grepreader,
  twoAppsSection: en.redesign.twoAppsSection,
  grepTagPage: en.redesign.grepTagPage,
};

// Keys that should NOT be translated (proper nouns, technical terms, brand names)
const SKIP_PATTERNS = ['iOS', 'macOS', 'Windows', 'EPUB', 'TXT', 'CBZ', 'PDF', 'Wi-Fi', 'AND', 'OR', 'NOT', 'App Store', 'BiblioFuse', 'GrepTag', 'iCloud', 'CJK'];

function shouldSkip(value) {
  if (typeof value !== 'string') return false;
  // Skip very short strings that are proper nouns or unchanged across languages
  if (value === value.toUpperCase() && value.length < 10) return true;
  return false;
}

function flattenKeys(obj, prefix = '') {
  const result = {};
  for (const [k, v] of Object.entries(obj)) {
    const key = prefix ? `${prefix}.${k}` : k;
    if (v && typeof v === 'object' && !Array.isArray(v)) {
      Object.assign(result, flattenKeys(v, key));
    } else {
      result[key] = v;
    }
  }
  return result;
}

function setNestedValue(obj, dotPath, value) {
  const parts = dotPath.split('.');
  let cur = obj;
  for (let i = 0; i < parts.length - 1; i++) {
    if (!cur[parts[i]]) cur[parts[i]] = {};
    cur = cur[parts[i]];
  }
  cur[parts[parts.length - 1]] = value;
}

async function translateForLang(langCode, langName) {
  const flat = flattenKeys(keysToTranslate);
  const toTranslate = {};
  for (const [k, v] of Object.entries(flat)) {
    if (typeof v === 'string' && v.trim() && !shouldSkip(v)) {
      toTranslate[k] = v;
    }
  }

  const prompt = `You are a professional translator for a mobile app marketing website. Translate the following JSON values from English to ${langName}.

Rules:
- Keep brand names unchanged: BiblioFuse, GrepTag, App Store, iCloud, Wi-Fi
- Keep technical abbreviations unchanged: EPUB, TXT, CBZ, PDF, CBR, RAR, ZIP, iOS, macOS, Windows, AND, OR, NOT, CJK
- Keep punctuation style (dashes, dots, arrows →) consistent
- Keep the same tone: direct, clear, marketing-friendly
- Translate only the values, keep the keys exactly as-is
- Return ONLY valid JSON with the same keys

${JSON.stringify(toTranslate, null, 2)}`;

  const msg = await client.messages.create({
    model: 'claude-haiku-4-5-20251001',
    max_tokens: 4096,
    messages: [{ role: 'user', content: prompt }],
  });

  const text = msg.content[0].text;
  const jsonMatch = text.match(/\{[\s\S]*\}/);
  if (!jsonMatch) throw new Error(`No JSON in response for ${langCode}`);

  return JSON.parse(jsonMatch[0]);
}

async function main() {
  for (const [langCode, langName] of Object.entries(LANGS)) {
    process.stdout.write(`Translating ${langCode} (${langName})... `);
    try {
      const translations = await translateForLang(langCode, langName);

      const filePath = join(localesDir, langCode, 'translation.json');
      const data = JSON.parse(readFileSync(filePath, 'utf8'));

      // Apply translations
      for (const [dotKey, value] of Object.entries(translations)) {
        // dotKey is like "grepreader.desc" or "twoAppsSection.eyebrow" or "grepTagPage.hero.title"
        // Map back to redesign.* namespace
        const fullKey = dotKey.startsWith('grepreader.')
          ? `redesign.productFamily.products.${dotKey}`
          : dotKey.startsWith('twoAppsSection.')
          ? `redesign.${dotKey}`
          : dotKey.startsWith('grepTagPage.')
          ? `redesign.${dotKey}`
          : null;

        if (fullKey) setNestedValue(data, fullKey, value);
      }

      writeFileSync(filePath, JSON.stringify(data, null, 2) + '\n', 'utf8');
      console.log('✅');
    } catch (e) {
      console.log(`❌ ${e.message}`);
    }
  }
  console.log('\n🎉 GrepTag key translations complete.');
}

main();

File: scripts/translate-grepreader-keys.py
import anthropic, json, sys, os

locales_dir = os.path.join(os.path.dirname(__file__), '..', 'src', 'locales')
client = anthropic.Anthropic()

LANGS = {
    'es': 'Spanish', 'fr': 'French', 'nl': 'Dutch', 'pt': 'Portuguese',
    'ru': 'Russian', 'zh': 'Simplified Chinese', 'ja': 'Japanese',
    'ko': 'Korean', 'id': 'Indonesian', 'ms': 'Malay'
}

with open(os.path.join(locales_dir, 'en', 'translation.json')) as f:
    en = json.load(f)

keys_to_translate = {
    'grepreader': en['redesign']['productFamily']['products']['grepreader'],
    'twoAppsSection': en['redesign']['twoAppsSection'],
    'grepTagPage': en['redesign']['grepTagPage'],
}

def flatten(obj, prefix=''):
    result = {}
    for k, v in obj.items():
        key = f"{prefix}.{k}" if prefix else k
        if isinstance(v, dict):
            result.update(flatten(v, key))
        else:
            result[key] = v
    return result

def set_nested(obj, dot_path, value):
    parts = dot_path.split('.')
    cur = obj
    for p in parts[:-1]:
        cur = cur.setdefault(p, {})
    cur[parts[-1]] = value

def translate_for_lang(lang_code, lang_name):
    flat = flatten(keys_to_translate)
    to_translate = {k: v for k, v in flat.items() if isinstance(v, str) and v.strip()}

    prompt = f"""Translate the following JSON values from English to {lang_name} for a mobile app marketing website.

Rules:
- Keep brand names unchanged: BiblioFuse, GrepTag, App Store, iCloud, Wi-Fi
- Keep technical abbreviations unchanged: EPUB, TXT, CBZ, PDF, CBR, RAR, ZIP, iOS, macOS, Windows, AND, OR, NOT, CJK
- Keep the same tone: direct, clear, marketing-friendly
- Translate only the values, keep the keys exactly as-is
- Return ONLY valid JSON with the same keys, no extra text

{json.dumps(to_translate, indent=2, ensure_ascii=False)}"""

    msg = client.messages.create(
        model='claude-haiku-4-5-20251001',
        max_tokens=4096,
        messages=[{'role': 'user', 'content': prompt}]
    )
    text = msg.content[0].text
    start = text.find('{')
    end = text.rfind('}') + 1
    return json.loads(text[start:end])

for lang_code, lang_name in LANGS.items():
    print(f"Translating {lang_code} ({lang_name})...", end=' ', flush=True)
    try:
        translations = translate_for_lang(lang_code, lang_name)
        file_path = os.path.join(locales_dir, lang_code, 'translation.json')
        with open(file_path) as f:
            data = json.load(f)

        for dot_key, value in translations.items():
            # Map to full redesign.* path
            if dot_key.startswith('grepreader.'):
                full_key = f"redesign.productFamily.products.{dot_key}"
            elif dot_key.startswith('twoAppsSection.') or dot_key.startswith('grepTagPage.'):
                full_key = f"redesign.{dot_key}"
            else:
                continue
            set_nested(data, full_key, value)

        with open(file_path, 'w') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
            f.write('\n')
        print('✅')
    except Exception as e:
        print(f'❌ {e}')

print('\n🎉 GrepTag key translations complete.')
